src/shared/index.ts

```diff
@@ -0,0 +1,214 @@
+
+			result.level = result.level ?? 'warning'
+			result.baselineState = result.baselineState ?? 'new'
+			result._suppression = !result.suppressions || result.suppressions.every(sup => sup.status === 'rejected')
+				? 'not suppressed'
+				: 'suppressed'
+		})
+
+		run._implicitBase = implicitBase?.join('/')
+		run.results?.forEach(result => {
+			result._relativeUri = result._uri?.replace(run._implicitBase ?? '' , '') ?? '' // For grouping, Empty works more predictably than undefined
+		})
+	})
+	log._distinct = mapDistinct(fileAndUris)
+	log._jsonMap = undefined // Free-up memory.
+}
+
+/*
+TfLoc
+   location: Loc
+
+Result
+   locations: Loc[]
+
+Loc
+   Message
+   PhyLoc
+      ArtLoc: Uri, Index
+      Region
+
+Run.artifacts: Art[]
+   location: ArtLoc
+   contents: ArtCon
+*/
+export function parseLocation(result: Result, loc?: Location) {
+	const message = loc?.message?.text
+	const [uri, uriContent] = parseArtifactLocation(result, loc?.physicalLocation?.artifactLocation)
+	const region = loc?.physicalLocation?.region
+	return { message, uri, uriContent, region }
+}
+
+export function parseRegion(region: Region | undefined): _Region | undefined {
+	if (!region) return undefined
+
+	const {byteOffset, byteLength} = region
+	if (byteOffset !== undefined && byteLength !== undefined) return [byteOffset, byteLength] as [number, number]
+
+	let {startLine, startColumn, endLine, endColumn} = region
+	if (!startLine) return undefined // Lines are 1-based so no need to check undef.
+
+	startLine--
+	if (!startColumn) return startLine
+
+	startColumn--
+	if (endColumn) endColumn--
+	if (endLine) endLine--
+	return [
+		startLine,
+		startColumn,
+		endLine ?? startLine,
+		endColumn ?? (startColumn + 1)
+	] as [number, number, number, number]
+}
+
+// Improve: `result` purely used for `_run.artifacts`.
+export function parseArtifactLocation(result: Result, anyArtLoc: ArtifactLocation | undefined) {
+	if (!anyArtLoc) return [undefined, undefined]
+	const runArt = result._run.artifacts?.[anyArtLoc.index ?? -1]
+	const runArtLoc = runArt?.location
+	const runArtCon = runArt?.contents
+
+	const uri = anyArtLoc.uri ?? runArtLoc?.uri // If index (§3.4.5) is absent, uri SHALL be present.
+	const uriContents = runArtCon?.text || runArtCon?.binary
+			? encodeURI(`sarif:${encodeURIComponent(result._log._uri)}/${result._run._index}/${anyArtLoc.index}/${uri?.file ?? 'Untitled'}`)
+			: undefined
+	return [uri, uriContents]
+}
+
+export const filtersRow = {
+	Level: {
+		'Error': true,
+		'Warning': true,
+		'Note': true,
+		'None': true,
+	},
+	Baseline: {
+		'New': true,
+		'Unchanged': true,
+		'Updated': true,
+		'Absent': false,
+	},
+	Suppression: {
+		'Not Suppressed': true,
+		'Suppressed': false,
+	},
+} as Record<string, Record<string, boolean>>
+
+export const filtersColumn = {
+	Columns: {
+		'Baseline': false,
+		'Suppression': false,
+		'Rule': false,
+	},
+} as Record<string, Record<string, boolean>>
```

Comment:
>as Record<string, Record<string, boolean>> [](start = 2, length = 42)

Instead of having `as TYPE` would specifying the type after declaration achieve the same thing? If so that's the convention we've been using, and we would prefer that

Reply:
Changed. I have been pretty consistent with the "as TYPE"-style. I would need time to change every instance over to the ": TYPE" style.

---
In reply to: [434795530](https://github.com/microsoft/sarif-vscode-extension/pull/279#discussion_r434795530) [](ancestors = 434795530)